File: Week_09/java/src/_125_ValidPalindrome.java
/**
 * 第一遍：2020/07/17周五 ✅
 * 第二遍：2020/07/18周六 ✅
 * 第三遍：2020/07/08周四
 * 第四遍：2020/07/13周一
 * 第三遍：2020/06/29周一
 * 第四遍：2020/07/05周日
 */
class _125_ValidPalindrome {
    public boolean isPalindrome(String s) {
        if (s == null || s.length() == 0) return true;
        int l = 0, r = s.length() - 1;
        while (l < r) {
            while (l < r && !Character.isLetterOrDigit(s.charAt(l))) l++;
            while (l < r && !Character.isLetterOrDigit(s.charAt(r))) r--;
            if (Character.toLowerCase(s.charAt(l)) == Character.toLowerCase(s.charAt(r))) {
                l++;
                r--;
            } else {
                return false;
            }
        }
        return true;
    }
}

File: Week_09/kotlin/src/_125_ValidPalindrome.kt
fun isPalindrome(s: String): Boolean {
    if (s.isEmpty()) return true
    var l = 0
    var r = s.length - 1
    while (l < r) {
        while (l < r && !Character.isLetterOrDigit(s[l])) l++
        while (l < r && !Character.isLetterOrDigit(s[r])) r--
        if (Character.toLowerCase(s[l]) == Character.toLowerCase(s[r])) {
            l++
            r--
        } else {
            return false
        }
    }
    return true
}

fun main() {
    println(isPalindrome("A man, a plan, a canal: Panama"))
}